packages/opentelemetry-plugin-ioredis/src/utils.ts

```diff
@@ -0,0 +1,92 @@
+/*!
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as ioredisTypes from 'ioredis';
+import { Tracer, SpanKind, Span, CanonicalCode } from '@opentelemetry/types';
+import { IORedisPluginClientTypes, IORedisCommand } from './types';
+import { IORedisPlugin } from './ioredis';
+import { AttributeNames } from './enums';
+
+const endSpan = (span: Span, err?: Error | null) => {
+  if (err) {
+    span.setStatus({
+      code: CanonicalCode.UNKNOWN,
+      message: err.message,
+    });
+  } else {
+    span.setStatus({ code: CanonicalCode.OK });
+  }
+  span.end();
+};
+
+export const getTracedSendCommand = (tracer: Tracer, original: Function) => {
+  return function sendCommandTrace(
+    this: ioredisTypes.Redis & IORedisPluginClientTypes,
+    cmd?: IORedisCommand
+  ) {
+    const parentSpan = tracer.getCurrentSpan();
+
+    if (arguments.length === 1 && typeof cmd === 'object') {
+      const span = tracer.startSpan(`${IORedisPlugin.COMPONENT}-${cmd.name}`, {
+        kind: SpanKind.CLIENT,
+        parent: parentSpan || undefined,
+        attributes: {
+          [AttributeNames.COMPONENT]: IORedisPlugin.COMPONENT,
+          [AttributeNames.DB_STATEMENT]: cmd.name,
+        },
+      });
+
+      // Set attributes for not explicitly typed IORedisPluginClientTypes
+      if (this.options) {
+        span.setAttributes({
+          [AttributeNames.PEER_HOSTNAME]: this.options.host,
+          [AttributeNames.PEER_PORT]: this.options.port,
+        });
+      }
+      if (this.address) {
+        span.setAttribute(
+          AttributeNames.PEER_ADDRESS,
+          `redis://${this.address}`
+        );
+      }
+
+      const command = arguments[0];
+      const originalCallback = command.callback;
+      if (originalCallback) {
+        (command as IORedisCommand).callback = function callback<T>(
+          this: unknown,
+          err: Error | null,
+          _reply: T
+        ) {
+          endSpan(span, err);
+          return originalCallback.apply(this, arguments);
+        };
+      } else {
+        console.error('Cannot figure out your callback', command);
```

Comment:
Use tracer logger isntead, although I'm not sure this line could ever be hit anyways. AFAIK redis/ioredis module will add a noop callback if the user doesn't provide one, but I could be wrong... Do any of your tests hit this line?

Reply:
Hi @markwolff currently they all do. `command.callback` is always null

Is there a way to use something like `command.promise`/`originalPromise` instead of `command.callback`/`originalCallback` ?

Reply:
It seems a promise a returned if no callback is provided. For these scenarios, you should do something like the postgres plugin, where you create a "passthrough" Promise. When the original promise resolves/rejects, it ends the span and then proceeds to the user's Promise handling logic.
https://github.com/open-telemetry/opentelemetry-js/blob/df2cc4347ec895b8d5d005f759c4293f03b93c40/packages/opentelemetry-plugin-postgres/opentelemetry-plugin-pg/src/pg.ts#L133-L153

Reply:
Indeed the main use case for `ioredis` is now to use `async/await` so we should handle that case like the OC plugin did.

Reply:
This is the part I am having most trouble with, I want to imitate [what was done in OC](https://github.com/census-instrumentation/opencensus-node/blob/ef5712fd3b279b0e80494322231232047b06f9e6/packages/opencensus-instrumentation-ioredis/src/ioredis.ts#L97-L115) as much as possible but need to understand what OpenTelemetry's equivalents are for [plugin.tracer.wrap](https://github.com/census-instrumentation/opencensus-node/blob/3014a954ce98c9ea0168d8a54390a63eb2b1e6d2/packages/opencensus-core/src/trace/model/tracer.ts#L137) and  [plugin.patchEnd](https://github.com/census-instrumentation/opencensus-node/blob/3014a954ce98c9ea0168d8a54390a63eb2b1e6d2/packages/opencensus-instrumentation-ioredis/src/ioredis.ts#L126)

Please stay tuned...

Reply:
I've tried adding the pg promise logic as per suggested when there is no cb

Reply:
Currently, async/await ioredis calls' spans are created properly. But not when used in callback form. Any ideas?